add auto-functionalize support for mutable list[Tensor]

bdhirsh opened the pull request:
Adds support for handling mutable List[Tensor] in auto-functionalization (@yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127347
* #128331
* #128045
* #128570
* #127927



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

bdhirsh committed:
add auto-functionalize support for mutable list[Tensor]

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_misc.py:
```diff
@@ -828,6 +828,48 @@ def forward(self, arg0_1: "f32[3]", arg1_1: "f32[3]", arg2_1: "f32[3]", arg3_1:
+
+            compiled_args = pytree.tree_map_only(torch.Tensor, torch.clone, orig_args)
+            with ctx():
+                torch.compile(f, backend="aot_eager", fullgraph=True)(
```
What happens with backend = inductor? Does it raise an error? If so, what does it look like?

bdhirsh replied:
the op as I wrote it (borrowed from @yf225) has an out= kwarg, so inductor fails on an assert that it doesn't support out= https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L290

I just tried making the op positional-only, and we fail a different assert here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L5314

I originally figured I would not bother biting off out= handling in inductor, so I would leave inductor alone. But I guess that's an independent issue from mutable list[Tensor] handling in inductor.  Let me know what you think - if you think it's worth getting mutable list[tensor] handling working in inductor, happy to do it as part of this PR.

(my tentative thought at first was that it wasn't worth "fixing" in inductor unless we also do the work of properlying getting list[tensor] support in inductor's reinplacing pass, since we probably never want the functional op to show up at runtime, but maybe they're worth tackling separately)

zou3519 replied:
> I just tried making the op positional-only, and we fail a different assert here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L5314

Can we just make the error message here better? ("You might be here if you have an op that has a mutable List[Tensor], we don't support that yet; file an issue"). 

> if you think it's worth getting mutable list[tensor] handling working in inductor, happy to do it as part of this PR.

Nah seems like too much work for now

bdhirsh replied:
sg, made the error message in inductor a bit better

zou3519 commented on test/dynamo/test_misc.py:
```diff
@@ -828,6 +828,48 @@ def forward(self, arg0_1: "f32[3]", arg1_1: "f32[3]", arg2_1: "f32[3]", arg3_1:
 
+    def test_auto_functionalize_tensorlist(self):
+        try:
+            lib = torch.library.Library("mylib", "FRAGMENT")
```
use torch._library.scoped_library in tests

zou3519 replied:
cc @kit1980 does torchlint get triggered on PRs?

kit1980 replied:
TorchFix should trigger and fail this, let me doublecheck.

kit1980 replied:
@zou3519 all flake8 lints are ignored for the file because line 7285 has "flake8: noqa".
It doesn't work as somebody who added that thought it works, it just unconditionally excludes the whole file from the linter.

zou3519 replied:
thanks for pointing that out

kit1980 replied:
Created an issue https://github.com/pytorch/pytorch/issues/127352

bdhirsh replied:
updated

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -92,6 +92,11 @@ def can_auto_functionalize(op: torch._ops.OperatorBase) -> bool:
+            type(arg.type) is torch.ListType
+            and type(arg.type.getElementType()) is torch.TensorType
+        ):
+            continue
```
Can you change the comment below? It looks like we now support Tensor[]

bdhirsh replied:
done

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -244,11 +251,26 @@ def do_auto_functionalize(
+            orig_args = normalized_kwargs[name]
+            assert len(orig_args) == len(unwrapped_out)
+            for orig_arg, o in zip(orig_args, unwrapped_out):
+                ctx.replace(orig_arg, o)
+                ctx.commit_update(orig_arg)
+                ctx.sync(orig_arg)
```
nit: maybe put into helper function?

bdhirsh replied:
done

zou3519 reviewed (commented):
Code looks good, I want to check that the inductor error message (if it exists) is sane

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

zou3519 commented on torch/_inductor/ir.py:
```diff
@@ -5246,6 +5249,9 @@ def handle_aliasing_and_mutation(info, arg):
                 return
             # can_auto_functionalize already filters out mutable List[Tensor].
             # We can support this in the future, but this is very uncommon.
+            assert (
+                not is_list_tensor
+            ), "Operators that accept mutable List[Tensor] are not yet supported in inductor"
```
Can you add a test that shows that this is the error one gets if they try backend="inductor"?

zou3519 reviewed (approved):
LGTM, but please add a test for the assertion (and CI looks like it's failing)

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "add auto-functionalize support for mutable list[Tensor]"

Adds support for handling mutable List[Tensor] in auto-functionalization (yf225 needs it for fsdp).

Note that this PR does not do the work of updating the reinplacing pass in inductor to properly remove the AutoFunctionalize op




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]
